src/tstool/tstool.c

```diff
@@ -0,0 +1,620 @@
+		format_flags(rx_filter_str, sizeof rx_filters_str, rx_filters, NUM_RX_FILTERS, 1 << so_ts_req.rx_filter);
+		printf("interface: %s\n"
+		       "clock: %s\n"
+		       "mac-address: %s\n"
+		       "fw-version: %s\n"
+		       "supported-apis:%s%s\n"
+		       "supported-tx-modes:%s\n"
+		       "supported-rx-filters:%s\n"
+		       "supported-sof-flags:%s\n"
+		       "tx-mode:%s\n"
+		       "rx-filter:%s\n",
+		       sfptpd_interface_get_name(intf),
+		       clock == NULL ? "none" : sfptpd_clock_get_short_name(clock),
+		       sfptpd_interface_get_mac_string(intf),
+		       sfptpd_interface_get_fw_version(intf),
+		       supports_phc ? " phc" : "",
+		       supports_efx ? " efx" : "",
+		       tx_types_str,
+		       rx_filters_str,
+		       sof_str,
+		       tx_type_str,
+		       rx_filter_str);
+		break;
+	case INTF_CMD_SET_TS:
+		intf = interfaces[0];
+		if (argc < 3 + cmd->intf_args) {
+			ERROR("insufficient arguments for timestamp configuration\n");
+			return EXIT_FAILURE;
+		}
+		if (decode_option(tx_types, NUM_TX_TYPES, &tx, argv[1 + cmd->intf_args]))
+			return EXIT_FAILURE;
+		if (decode_option(rx_filters, NUM_RX_FILTERS, &rx, argv[2 + cmd->intf_args]))
+			return EXIT_FAILURE;
+		so_ts_req.tx_type = tx;
+		so_ts_req.rx_filter = rx;
+		rc = sfptpd_interface_ioctl(intf, SIOCSHWTSTAMP, &so_ts_req);
+		break;
+	default:
+		fprintf(stderr, "unknown interface command: %s\n", command);
+		usage(stderr);
+		return EXIT_FAILURE;
+	}
+
+	if (rc != 0) {
+		ERROR("tstool: interface: %s: %s\n", command, strerror(rc));
+		return EXIT_FAILURE;
+	}
+
+	return EXIT_SUCCESS;
+}
+
+
+/****************************************************************************
+ * Global functions
+ ****************************************************************************/
+
+int main(int argc, char *argv[])
+{
+        struct sfptpd_config_general *gconf;
+	const char *subsystem;
+	int index;
+	int opt;
+	int rc = EXIT_FAILURE;
+	int ret;
+
+	/* Initialise config */
+	ret = sfptpd_config_create(&config);
+	if (ret != 0)
+		return EXIT_FAILURE;
+
+	/* Tweak config */
+	gconf = sfptpd_general_config_get(config);
+	gconf->non_sfc_nics = true;
+	gconf->timestamping.disable_on_exit = false;
+	gconf->clocks.persistent_correction = false;
+	gconf->clocks.no_initial_correction = true;
+
+	/* Handle command line arguments */
+	while ((opt = getopt_long(argc, argv, opts_short, opts_long, &index)) != -1) {
+		switch (opt) {
+		case 'h':
+			usage(stdout);
+			goto fail;
+		case 'v':
+			sfptpd_log_set_trace_level(SFPTPD_COMPONENT_ID_NETLINK, 3);
+			sfptpd_log_set_trace_level(SFPTPD_COMPONENT_ID_SFPTPD, 6);
+			sfptpd_log_set_trace_level(SFPTPD_COMPONENT_ID_CLOCKS, 3);
+			break;
+		case OPT_PERSISTENT:
+			gconf->clocks.persistent_correction = true;
+			break;
+		case OPT_INITIAL:
+			gconf->clocks.no_initial_correction = false;
+			break;
+		default:
+			fprintf(stderr, "unexpected option: %s\n", argv[optind]);
```

Comment:
Would it be better to default to "safe" behaviour i.e. read-only commands run immediately, but we print info and ask for confirmation before altering any settings?

Could have a `force`/`y`/`always` option to allow scripts to run "destructive" commands without the prompt.

Reply:
I'm not massively keen on that idea - I can't imagine the `dedup` command being used interactively on production systems without thought. Maybe I'm wrong.